### PR TITLE
Subdelegations, airdrop example is now a subdelegation example, up to 30% less gas for delegate methods

### PR DESCRIPTION
**Pull request by mireynolds**
### Replaced OpenZeppelin enumerable sets with `bytes32[]` array in `vault` and `delegate` hash mappings ###

Hashes for a particular `vault` / `delegate` are now pushed to a `bytes32[]` array in their hash mappings instead of adding (and removing) them from a set. This reduces gas by > 40k every time a delegation is added to the registry due to the storage writes in the enumerable set backend. Suppose that a `vault` / `delegate` accumulates 500 delegation hashes in their array, the increased computation of the `getDelegationsFor` methods is well within acceptable gas limits for most RPC providers.

Consumable methods are unaffected as they now use the `_delegations` mapping only to establish whether a delegation is valid and enabled / disabled.

### Changed `bytes` to `bytes32[6]` in `_delegations` mapping ###

Reduces > 20k gas every time a delegation is added since variable sized storage data also need to store their length.

### Updated delegate methods to write minimal storage and removed `_setDelegationValues` ###

Delegate methods now only write data if it exists in their type and does not write `rights` to storage if it's the default `""`. `_computeDelegationLocation(hash)` is used to obtain the storage location for a given hash, and `writeDelegation(location, StoragePositions.delegate, data);` is used to write particular delegation data. For example, this saves > 80k gas for a `delegateForAll` call with all rights. Storage positions of data types are standardised with the new `StoragePositions` struct. `_setDelegationValues` has been removed since these changes make it redundant.

### Delegation type now encoded in the last byte of the hash ###

This solves both the problem of collisions and being able to identify the type of delegation without writing its type to storage.

### Check methods now support subdelegations ###

All check methods now accept a `bytes32 rights` parameter (previously `data`). If this parameter is not the default `""`, the check will bubble through the default case and then the `rights` case unless it returns `true` before exhausting the search (in the case of erc721, contract, all) or unless it returns `type(uint256).max ` before exhausting the search (in the erc20, erc1155 cases). The `_loadDelegationAddress(location, StoragePositions.vault)` and  `_loadDelegationUint(location, StoragePositions.balance)` methods are used to load `vault` and `balance` values directly from storage. A check on `vault` is now used to determine if the hash exists and is valid (since it is set to zero when disabled, and set to `vault` when enabled).

 ### Removed occurrences of `IDelegationRegistry.[object]`  ###

These objects are inherited explicitly anyway, and omitting `IDelegationRegistry` from these object references increases readability of the code.

### Created Consumable and Private methods sections ###

Mostly a code refactor to keep the write and read sections clean. Added consumable as a section since we have distinct classes of on chain functions that can be consumed on and off chain.

## IDelegateRegistry.sol changes ##

- Removed `NONE` from `DelegationType` enum as it's never used.
- Added `StoragePositions` enum to standardise storage position of delegation data.
- Renamed `DelegationInfo` to `Delegation` struct to be used for returning an arbitrary delegation.
- Created `BatchDelegation` struct to be used for batch delegations

## Airdrop Example Updated to a Subdelegation Example ##

The airdrop example now accepts sub delegations with `rights = "airdrop"` in addition to default delegations, configured with a `acceptableRight` variable in the airdrop contract.

## New tests ##

- Fuzz test for `supportsInterface` method
- `testVultEnumerationGas` tests an edge case where a vault has enabled 500 delegations.

### CODE REVIEW

File: src/IDelegateRegistry.sol [hunk 1]
```diff
@@ -10,40 +10,61 @@ pragma solidity ^0.8.20;
 interface IDelegateRegistry {
     /// @notice Delegation type
     enum DelegationType {
-        NONE,
```
Comment (0xfoobar) [inline src/IDelegateRegistry.sol:13]: Good to populate the zero enum with an empty NONE value, to avoid people passing incorrect default values

File: src/IDelegateRegistry.sol [hunk 2]
```diff
@@ -10,40 +10,61 @@ pragma solidity ^0.8.20;
...
         ALL,
         CONTRACT,
-        ERC721,
         ERC20,
+        ERC721,
```
Comment (0xfoobar) [inline src/IDelegateRegistry.sol:16]: Makes sense from an EIP ordering perspective, but wonder if it makes sense to preserve the same enum ordering from v1 here

File: src/IDelegateRegistry.sol [hunk 3]
```diff
@@ -10,40 +10,61 @@ pragma solidity ^0.8.20;
...
+        address delegate;
         address vault;
+        bytes32 rights;
+        address contract_;
+        uint256 tokenId;
+        uint256 balance;
+    }
+
+    /// @notice Struct for batch delegations
+    struct BatchDelegation {
```
Comment (0xfoobar) [inline src/IDelegateRegistry.sol:43]: Slightly slimmer to not have the vault address in here, less calldata in batchDelegate(), but does leave us with two annoying mostly parallel structs

File: src/DelegateRegistry.sol [hunk 4]
```diff
@@ -104,199 +100,375 @@ contract DelegateRegistry is IDelegateRegistry {
...
+        return DelegationType(uint8(uint256(_input) & 0xFF));
+    }
+
+    /// @dev Helper function that filters an array of delegation hashes by removing disabled delegations and duplicates
+    function _filterDelegationHashes(bytes32[] memory array_) private view returns (bytes32[] memory) {
+        bool duplicate;
+        uint256 count = 0;
+        bytes32[] memory tempArray = new bytes32[](array_.length);
+
+        for (uint256 i = 0; i < array_.length; i++) {
```
Comment (0xfoobar) [inline src/DelegateRegistry.sol:424]: We get quadratic complexity in the length of the hashes here, could be a real problem for Liquid Delegates and other extensions as they scale, would be 500k operations at array length 1k and 12 million operations at array length 5k

File: src/DelegateRegistry.sol [hunk 5]
```diff
@@ -104,199 +100,375 @@ contract DelegateRegistry is IDelegateRegistry {
...
     }
 
-    /// @dev Helper function to compute delegation hash for token delegation
-    function _computeDelegationHashForERC721(address vault, address delegate, address contract_, uint256 tokenId, bytes32 data)
-        internal
-        pure
-        returns (bytes32)
+    /// @inheritdoc IDelegateRegistry
+    function checkDelegateForERC721(address delegate, address vault, address contract_, uint256 tokenId, bytes32 rights)
+        public
```
Comment (0xfoobar) [inline src/DelegateRegistry.sol:274]: Can be made `external`

File: src/DelegateRegistry.sol [hunk 6]
```diff
@@ -104,199 +100,375 @@ contract DelegateRegistry is IDelegateRegistry {
...
+            location = _computeDelegationLocation(_computeDelegationHashForContract(contract_, delegate, rights, vault));
+            valid = checkDelegateForAll(delegate, vault, rights) || _loadDelegationAddress(location, StoragePositions.vault) == vault;
+        }
     }
 
-    /// @dev Helper function to compute delegation hash for contract delegation
-    function _computeDelegationHashForContract(address vault, address delegate, address contract_, bytes32 data) internal pure returns (bytes32) {
-        return keccak256(abi.encode(delegate, vault, contract_, data));
+    /// @inheritdoc IDelegateRegistry
+    function checkDelegateForERC20(address delegate, address vault, address contract_, bytes32 rights) external view override returns (uint256 balance) {
```
Comment (0xfoobar) [inline src/DelegateRegistry.sol:258]: Still thinking about whether this (& check1155) should diverge from the others and return a uint256, or expect the `amount` as a param and simply return a `bool` validity result

File: src/IDelegateRegistry.sol [hunk 7]
```diff
@@ -10,40 +10,61 @@ pragma solidity ^0.8.20;
...
+        rights,
+        contract_,
+        tokenId,
+        balance
+    }
+
+    /// @notice Struct for returning arbitrary delegations
+    struct Delegation {
         DelegationType type_;
+        bool enable;
```
Comment (0xfoobar) [inline src/IDelegateRegistry.sol:33]: We only want to return the `enabled` delegations, if false then it should be cleared from storage altogether. Think this struct is used primarily on the enumeration front, maybe a way to simplify by removing this boolean

File: src/IDelegateRegistry.sol (hunk 2 above)
Comment (mireynolds) [reply src/IDelegateRegistry.sol:17]: Happy to go either way. Think we can get probably get away with the EIP ordering since we're changing the name from `TOKEN` in v1 to `ERC721` in v2

File: src/IDelegateRegistry.sol (hunk 3 above)
Comment (mireynolds) [reply src/IDelegateRegistry.sol:44]: Honestly the additional call data doesn't bother me too much. More the potential discrepancy between `vault` in the calldata and `msg.sender` that is actually used

File: src/IDelegateRegistry.sol (hunk 7 above)
Comment (mireynolds) [reply src/IDelegateRegistry.sol:34]: `enabled` is never stored in storage. All this struct is used for is a structure in memory for returning arbitrary delegations in the `getDelegationsFor` functions.. which should all return as valid anyway. So maybe not needed :)

File: src/DelegateRegistry.sol (hunk 4 above)
Comment (mireynolds) [reply src/DelegateRegistry.sol:424]: Latest push makes it impossible to push duplicates, and have removed the duplicates check from the delegations filter! :)

File: src/DelegateRegistry.sol [hunk 8]
```diff
@@ -104,199 +100,367 @@ contract DelegateRegistry is IDelegateRegistry {
...
+        if (enable) {
+            if (_loadDelegationBytes32(location, StoragePositions.vault) == "") _pushDelegationHashes(msg.sender, delegate, hash);
+            _writeDelegation(location, StoragePositions.contract_, contract_);
+            _writeDelegation(location, StoragePositions.delegate, delegate);
+            _writeDelegation(location, StoragePositions.vault, msg.sender);
+            if (rights != "") _writeDelegation(location, StoragePositions.rights, rights);
+        } else {
+            _writeDelegation(location, StoragePositions.contract_, "");
+            _writeDelegation(location, StoragePositions.delegate, "");
+            _writeDelegation(location, StoragePositions.vault, 1);
```
Comment (0xfoobar) [inline src/DelegateRegistry.sol:157]: This messes with other getter functions now. On line 333, the `enable` param of getDelegationsFromHashes()` needs to be updated. And we should return a completely empty struct if the delegate doesn't exist

File: src/IDelegateRegistry.sol [hunk 9]
```diff
@@ -13,37 +13,59 @@ interface IDelegateRegistry {
...
-        ERC721,
         ERC20,
+        ERC721,
         ERC1155
     }
 
-    /// @notice Info about a single delegation, used for onchain enumeration
-    struct DelegationInfo {
+    /// @notice Standardizes storage positions of delegation data
+    enum StoragePositions {
```
Comment (0xfoobar) [inline src/IDelegateRegistry.sol:22]: Since this struct isn't user-facing and just an implementation detail, should we move its declaration out of the consumable interface and into the registry itself?

File: src/IDelegateRegistry.sol (hunk 2 above)
Comment (mireynolds) [reply src/IDelegateRegistry.sol]: Reverted to the v1 ordering, made ordering consistent in the interface and registry.
